rtCommon/bidsArchive.py

```diff
@@ -0,0 +1,825 @@
+        creating new directories if necessary (this behavior can be overridden).
+
+        Args:
+            incremental: BIDS Incremental to append
+            makePath: Create new directory path for BIDS-I data if needed.
+                (default: True).
+            validateAppend: Compares image metadata and NIfTI headers to check
+                that the images being appended are part of the same sequence and
+                don't conflict with each other (default: True).
+
+        Raises:
+            RuntimeError: If the image to append to in the archive is not either
+                3D or 4D.
+            StateError: If the image path within the BIDS-I would result in
+                directory creation and makePath is set to False.
+            ValidationError: If the data to append is incompatible with existing
+                data in the archive.
+
+        Returns:
+            True if the append succeeded, False otherwise.
+
+        Examples:
+            Assume we have a NIfTI image 'image' and a metadata dictionary
+            'metdata' with all required metadata for a BIDS Incremental.
+
+            >>> archive = BidsArchive('.')
+            >>> incremental = BidsIncremental(image, metadata)
+            >>> archive.appendIncremental(incremental)
+
+            If we don't want to create any new files/directories in the archive,
+            makePath can be set to false.
+
+            >>> archive = BidsArchive('/tmp/emptyDirectory')
+            >>> archive.appendIncremental(incremental, makePath=False)
+            False
+        """
+        # 1) Create target path for image in archive
+        dataDirPath = incremental.dataDirPath
+        imgPath = incremental.imageFilePath
+        metadataPath = incremental.metadataFilePath
+
+        # 2) Verify we have a valid way to append the image to the archive.
+        # 3 cases:
+        # 2.0) Archive is empty and must be created
+        # 2.1) Image already exists within archive, append this Nifti to it
+        # 2.2) Image doesn't exist in archive, but rest of the path is valid for
+        # the archive; create new Nifti file within the archive
+        # 2.3) Neither image nor path is valid for provided archive; fail append
+        if self.isEmpty() and makePath:
+            incremental.writeToDisk(self.rootPath)
+            self._updateLayout()
+
+        elif self.pathExists(imgPath):
+            logger.debug("Image exists in archive, appending")
+
+            # Supplement entity dict with file name entities that the BIDSLayout
+            # returns that aren't official bids entities
+            entityDict = incremental.entities
+            entityDict['datatype'] = incremental.datatype
+            entityDict['suffix'] = incremental.suffix
+
+            # The one exact match must exist because the path exists
+            images = self.getImages(**entityDict, matchExact=True)
+            assert len(images) == 1
```

Comment:
GW comment: I'm wondering if this assert should instead be a raise error. If len(images) !=1 is it a bug in the code or an error in the data or parameters?

Reply:
TODO: Add to assert message that this shouldn't happen because of definition of matchExact